radsat-sk/operation/subsystems/RPdb.c

```diff
@@ -0,0 +1,272 @@
+
+
+/**
+ * Constant Values for I2C Communication
+ */
+/** I2C Slave Address for PDB */
+#define PDB_I2C_SLAVE_ADDR 			(0x2B)
+/** Data Sent to the PDB via I2C for telemetry commands is 6 Bytes, 2 for the command, 4 for data*/
+#define PDB_TELEM_COMMAND_LENGTH	(6)
+/** Data sent to the PDB via I2C for all other commands is 4 Bytes, 2 for the command, 2 for data*/
+#define PDB_COMMAND_LENGTH			(4)
+/** Data Returned from the PDB via I2C is 4 Bytes */
+#define PDB_RESPONSE_LENGTH			(4)
+/** Delay between telemetry read/write operations with the PDB on I2C is 5ms */
+#define PDB_I2C_TELEM_DELAY			(5)
+/** Delay between other read/write operations with the PDB on I2C is 1ms */
+#define PDB_I2C_DELAY				(1)
+
+
+/**
+ * Constant Values for I2C Communication
+ */
+/** Number of sun sensors on the CubeSat */
+#define NUM_SUN_SENSORS		(6)
+
+
+
+
+/***************************************************************************************************
+                                          PRIVATE GLOBALS
+***************************************************************************************************/
+/**
+ *  4 Byte commands to request data for each of the 6 sun-sensors
+ *	located on each face of the CubeSat Solar Arrays.
+ */
+static uint32_t PdbSunSensorCommandBytes[6] = {
+	0x10E11C,     // SA1A - xPos? TODO: Fill out which side is which connection
+	0x10E11D,     // SA1B -
+	0x10E12C,     // SA2A -
+	0x10E12D,     // SA2B -
+	0x10E13C,     // SA3A -
+	0x10E13D      // SA3B -
+};
+
+/**
+ * 4 Byte commands for requesting output voltage readings for each bus and BCR
+ */
+static uint32_t PdbVoltageCommandBytes[4] = {
+		0x10E280,		// Output Voltage of BCR
+		0x10E220,		// Output Voltage of Battery Bus
+		0x10E210,		// Output Voltage of 5V Bus
+		0x10E200		// Output Voltage of 3.3V Bus
+};
+
+/**
+ * 4 Byte commands for requesting output current readings for each bus and BCR
+ */
+static uint32_t PdbCurrentCommandBytes[4] = {
+		0x10E284,		// Output Current of BCR in mA
+		0x10E224,		// Output Current of Battery Bus
+		0x10E214,		// Output Current of 5V Bus
+		0x10E204		// Output Current of 3.3V Bus
+};
+
+
+/***************************************************************************************************
+                                       PRIVATE FUNCTION STUBS
+***************************************************************************************************/
+
+static uint16_t PdbI2cTalk(uint32_t command);
+
+
+/***************************************************************************************************
+                                             PUBLIC API
+***************************************************************************************************/
+/**
+ * Get the Irradiance value from the sun sensor on each face of the CubeSat
+ *
+ * @return A pointer to a SunSensorStatus object containing the converted ADC data for each sun sensor
+ */
+SunSensorStatus getSunSensorData(void) {
```

Comment:
It would be a better approach to return a pointer to a struct as an output parameter and have the function return the error (if there is one) 

A good example on how to do this is in alpha RTransiecver.c, the transcieverTelemetry function